src/Microsoft.IdentityModel.Clients.ActiveDirectory/Internal/Instance/InstanceDiscovery.cs

```diff
@@ -0,0 +1,151 @@
+
+        private static HashSet<string> WhitelistedAuthorities = new HashSet<string>(new[]
+        {
+            "login.windows.net", // Microsoft Azure Worldwide - Used in validation scenarios where host is not this list 
+            "login.chinacloudapi.cn", // Microsoft Azure China
+            "login.microsoftonline.de", // Microsoft Azure Blackforest
+            "login-us.microsoftonline.com", // Microsoft Azure US Government - Legacy
+            "login.microsoftonline.us", // Microsoft Azure US Government
+            "login.microsoftonline.com" // Microsoft Azure Worldwide
+        });
+
+        public const string AuthorizeEndpointTemplate = "https://{host}/{tenant}/oauth2/authorize";
+
+        // The following cache could be private, but we keep it public so that internal unit test can take a peek into it
+        public static ConcurrentDictionary<string, InstanceDiscoveryMetadataEntry> InstanceCache =
+            new ConcurrentDictionary<string, InstanceDiscoveryMetadataEntry>(); // Keys are host strings
+
+        private static SemaphoreSlim semaphore = new SemaphoreSlim(1, 1);
+
+        public static async Task<InstanceDiscoveryMetadataEntry> GetMetadataEntry(string host, bool validateAuthority,
+            CallState callState)
+        {
+            InstanceDiscoveryMetadataEntry entry = null;
+            if (!InstanceCache.TryGetValue(host, out entry))
+            {
+                await semaphore.WaitAsync(); // SemaphoreSlim.WaitAsync() will not block current thread
+                try
+                {
+                    if (!InstanceCache.TryGetValue(host, out entry))
+                    {
+                        await DiscoverAsync(host, validateAuthority, callState);
+                        InstanceCache.TryGetValue(host, out entry);
+                    }
+                }
+                finally
+                {
+                    semaphore.Release();
+                }
+            }
+            return entry;
+        }
+
+        private static async Task DiscoverAsync(string host, bool validateAuthority,
+            CallState callState) // No return value. Modifies InstanceCache directly.
+        {
+            string tentativeAuthorizeEndpoint = AuthorizeEndpointTemplate.Replace("{host}", host);
+            string instanceDiscoveryEndpoint =
+                ("https://{host}/common/discovery/instance?api-version=1.1&authorization_endpoint=" +
+                 tentativeAuthorizeEndpoint)
+                .Replace("{host}", WhitelistedAuthorities.Contains(host) ? host : DefaultTrustedAuthority);
+            var client = new AdalHttpClient(instanceDiscoveryEndpoint, callState);
+            InstanceDiscoveryResponse discoveryResponse = null;
+            try
+            {
+                discoveryResponse = await client.GetResponseAsync<InstanceDiscoveryResponse>().ConfigureAwait(false);
+                if (validateAuthority & discoveryResponse.TenantDiscoveryEndpoint == null)
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Correct this '&' to '&&'. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://iddp-sonarqube:9000/coding_rules#rule_key=csharpsquid%3AS2178)